pkg/storage/repo/storage.go

```diff
@@ -0,0 +1,269 @@
+	_, span := system.GetTracer().Start(ctx, "pkg/storage/repo/repo.PrepareStorage")
+	defer span.End()
+	repoURL := storageSpec.Repo
+	var err error
+	_, err = clone.IsValidGitRepoURL(repoURL)
+	// fmt.Printf("%+v", storageSpec)
+	if err != nil {
+		return storage.StorageVolume{}, err
+	}
+	// # create a tmp directory
+	outputPath, err := os.MkdirTemp(sp.LocalDir, "*")
+	fmt.Printf("Output Path: %s\n", outputPath)
+	if err != nil {
+		return storage.StorageVolume{}, err
+	}
+
+	// err = git.Clone(repoURL, outputPath)
+	// if err != nil {
+	// 	fmt.Println("Failed to clone repository:", err)
+	cmd := exec.Command("git", "clone", repoURL, outputPath)
+	// The `Output` method executes the command and
+	// collects the output, returning its value
+	out, err1 := cmd.Output()
+	if err1 != nil {
+		// if there was any error, print it here
+		fmt.Println("could not run command: ", err)
+	}
+	// otherwise, print the output from running the command
+	fmt.Println("Output: ", string(out))
+	if err != nil {
+		panic(err)
+	}
+
+	if err != nil {
+		return storage.StorageVolume{}, err
+	}
+	// }
+	filepath, err2 := url.Parse(repoURL)
+	if err2 != nil {
+		return storage.StorageVolume{}, err
+	}
+	filename := strings.Split(filepath.Path, ".")[0]
+	targetPath := "/inputs" + filename
+
+	CIDSpec, err := sp.Upload(ctx, outputPath)
+	// If estuary key exists then upload it to estuary
+	if err != nil {
+		return storage.StorageVolume{}, err
+	}
+	CID := CIDSpec.CID
+	// Update the KV store
+	SHA1HASH, _ := UrltoLatestCommitHash(repoURL)
+	spkey := sp.EstuaryAPIKey
+	envkey := os.Getenv("ESTUARY_API_KEY")
+	var key string
+	if spkey != "" && envkey == "" {
+		key = spkey
+	} else if envkey != "" && spkey == "" {
+		key = envkey
+	} else if envkey != "" && spkey != "" {
+		key = spkey
+	} else {
+		key = ""
+	}
+	if key != "" {
+		log1.Println("Pinning to Estuary...")
+		//nolint:govet,noctx
+		err := estuary.PinToIPFSViaEstuary(ctx, key, CID)
+		if err != nil {
+			return storage.StorageVolume{}, err
+		}
+		log1.Println("Successfully Pinned to Estuary...")
+	}
+	data := url.Values{}
+	data.Set("key", SHA1HASH)
+	data.Set("value", CID)
+	if err != nil {
+		return storage.StorageVolume{}, err
+	}
+	err = CreateSHA1CIDPair(data)
+	if err != nil {
+		fmt.Printf("Failed to create SHA1CIDPair: %v", err)
+	}
+	// return the volume
+	volume := storage.StorageVolume{
+		Type:   storage.StorageVolumeConnectorBind,
+		Source: outputPath,
+		Target: targetPath,
+	}
+
+	return volume, nil
+}
+
+func (sp *StorageProvider) Upload(ctx context.Context, localPath string) (model.StorageSpec, error) {
+	ctx, span := system.GetTracer().Start(ctx, "storage/repo/apicopy.Upload")
+	defer span.End()
+
+	cid, err := sp.IPFSClient.Upload(ctx, localPath)
+	fmt.Print(cid)
+	if err != nil {
+		fmt.Print(err)
+		return model.StorageSpec{}, err
+	}
+	return model.StorageSpec{
+		StorageSource: model.StorageSourceIPFS,
+		CID:           cid.CID,
+	}, nil
```

Comment:
As I said in the previous comments, this class should not be uploading to IPFS, but should be pushing back to a git repo. If you don't want to implement that, just `panic`.

Reply:
If we don't upload to IPFS we won't get a CID, If no CID then we can't create the kv pair for subsequent mounting of the repo

Reply:
We will implement push functionality later